File: bindings/lua/Polycode/ResourcePool.lua
require "Polycode/EventDispatcher"

class "ResourcePool" (EventDispatcher)


function ResourcePool:__getvar(name)
	if name == "reloadResourcesOnModify" then
		return Polycode.ResourcePool_get_reloadResourcesOnModify(self.__ptr)
	elseif name == "dispatchChangeEvents" then
		return Polycode.ResourcePool_get_dispatchChangeEvents(self.__ptr)
	elseif name == "resourceSubscribers" then
		return Polycode.ResourcePool_get_resourceSubscribers(self.__ptr)
	elseif name == "deleteOnUnsubscribe" then
		return Polycode.ResourcePool_get_deleteOnUnsubscribe(self.__ptr)
	end
	if EventDispatcher["__getvar"] ~= nil then
		return EventDispatcher.__getvar(self, name)
	end
end

function ResourcePool:__setvar(name,value)
	if name == "reloadResourcesOnModify" then
		Polycode.ResourcePool_set_reloadResourcesOnModify(self.__ptr, value)
		return true
	elseif name == "dispatchChangeEvents" then
		Polycode.ResourcePool_set_dispatchChangeEvents(self.__ptr, value)
		return true
	elseif name == "resourceSubscribers" then
		Polycode.ResourcePool_set_resourceSubscribers(self.__ptr, value)
		return true
	elseif name == "deleteOnUnsubscribe" then
		Polycode.ResourcePool_set_deleteOnUnsubscribe(self.__ptr, value)
		return true
	end
	if EventDispatcher["__setvar"] ~= nil then
		return EventDispatcher.__setvar(self, name, value)
	else
		return false
	end
end
function ResourcePool:ResourcePool(...)
	local arg = {...}
	if type(arg[1]) == "table" and count(arg) == 1 then
		if ""..arg[1].__classname == "EventDispatcher" then
			self.__ptr = arg[1].__ptr
			return
		end
	end
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.ResourcePool(unpack(arg))
	end
end

function ResourcePool:setFallbackPool(pool)
	local retVal = Polycode.ResourcePool_setFallbackPool(self.__ptr, pool.__ptr)
end

function ResourcePool:addResource(resource)
	local retVal = Polycode.ResourcePool_addResource(self.__ptr, resource.__ptr)
end

function ResourcePool:removeResource(resource)
	local retVal = Polycode.ResourcePool_removeResource(self.__ptr, resource.__ptr)
end

function ResourcePool:hasResource(resource)
	local retVal = Polycode.ResourcePool_hasResource(self.__ptr, resource.__ptr)
	return retVal
end

function ResourcePool:loadResourcesFromFolder(folder, recursive)
	local retVal = Polycode.ResourcePool_loadResourcesFromFolder(self.__ptr, folder, recursive)
end

function ResourcePool:loadResource(path)
	local retVal = Polycode.ResourcePool_loadResource(self.__ptr, path)
	if retVal == nil then return nil end
	local __c = _G["Resource"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function ResourcePool:loadResourceWithName(path, name)
	local retVal = Polycode.ResourcePool_loadResourceWithName(self.__ptr, path, name)
	if retVal == nil then return nil end
	local __c = _G["Resource"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function ResourcePool:getResource(resourceType, resourceName)
	local retVal = Polycode.ResourcePool_getResource(self.__ptr, resourceType, resourceName)
	if retVal == nil then return nil end
	local __c = _G["Resource"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function ResourcePool:getName()
	local retVal =  Polycode.ResourcePool_getName(self.__ptr)
	return retVal
end

function ResourcePool:setName(name)
	local retVal = Polycode.ResourcePool_setName(self.__ptr, name)
end

function ResourcePool:getResourceByPath(resourcePath)
	local retVal = Polycode.ResourcePool_getResourceByPath(self.__ptr, resourcePath)
	if retVal == nil then return nil end
	local __c = _G["Resource"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function ResourcePool:Update(elapsed)
	local retVal = Polycode.ResourcePool_Update(self.__ptr, elapsed)
end

function ResourcePool:getResources(resourceType)
	local retVal = Polycode.ResourcePool_getResources(self.__ptr, resourceType)
	if retVal == nil then return nil end
	for i=1,count(retVal) do
		local __c  = _G["Resource"]("__skip_ptr__")
		__c.__ptr = retVal[i]
		retVal[i] = __c
	end
	return retVal
end

function ResourcePool:checkForChangedFiles()
	local retVal =  Polycode.ResourcePool_checkForChangedFiles(self.__ptr)
end

function ResourcePool:__delete()
	if self then Polycode.delete_ResourcePool(self.__ptr) end
end

File: bindings/javascript/Polycode/ResourcePool.js
require('Polycode/EventDispatcher')

function ResourcePool() {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.ResourcePool()
	}
	Object.defineProperties(this, {
		'reloadResourcesOnModify': { enumerable: true, configurable: true, get: ResourcePool.prototype.__get_reloadResourcesOnModify, set: ResourcePool.prototype.__set_reloadResourcesOnModify},
		'dispatchChangeEvents': { enumerable: true, configurable: true, get: ResourcePool.prototype.__get_dispatchChangeEvents, set: ResourcePool.prototype.__set_dispatchChangeEvents},
		'resourceSubscribers': { enumerable: true, configurable: true, get: ResourcePool.prototype.__get_resourceSubscribers, set: ResourcePool.prototype.__set_resourceSubscribers},
		'deleteOnUnsubscribe': { enumerable: true, configurable: true, get: ResourcePool.prototype.__get_deleteOnUnsubscribe, set: ResourcePool.prototype.__set_deleteOnUnsubscribe}
	})
}


ResourcePool.prototype = Object.create(EventDispatcher.prototype)

ResourcePool.prototype.__get_reloadResourcesOnModify = function() {
	return Polycode.ResourcePool__get_reloadResourcesOnModify(this.__ptr)
}

ResourcePool.prototype.__set_reloadResourcesOnModify = function(val) {
	Polycode.ResourcePool__set_reloadResourcesOnModify(this.__ptr, val)
}

ResourcePool.prototype.__get_dispatchChangeEvents = function() {
	return Polycode.ResourcePool__get_dispatchChangeEvents(this.__ptr)
}

ResourcePool.prototype.__set_dispatchChangeEvents = function(val) {
	Polycode.ResourcePool__set_dispatchChangeEvents(this.__ptr, val)
}

ResourcePool.prototype.__get_resourceSubscribers = function() {
	return Polycode.ResourcePool__get_resourceSubscribers(this.__ptr)
}

ResourcePool.prototype.__set_resourceSubscribers = function(val) {
	Polycode.ResourcePool__set_resourceSubscribers(this.__ptr, val)
}

ResourcePool.prototype.__get_deleteOnUnsubscribe = function() {
	return Polycode.ResourcePool__get_deleteOnUnsubscribe(this.__ptr)
}

ResourcePool.prototype.__set_deleteOnUnsubscribe = function(val) {
	Polycode.ResourcePool__set_deleteOnUnsubscribe(this.__ptr, val)
}

Duktape.fin(ResourcePool.prototype, function (x) {
	if (x === ResourcePool.prototype) {
		return;
	}
	Polycode.ResourcePool__delete(x.__ptr)
})

ResourcePool.prototype.addResource = function(resource) {
	Polycode.ResourcePool_addResource(this.__ptr, resource)
}

ResourcePool.prototype.removeResource = function(resource) {
	Polycode.ResourcePool_removeResource(this.__ptr, resource)
}

ResourcePool.prototype.hasResource = function(resource) {
	return Polycode.ResourcePool_hasResource(this.__ptr, resource)
}

ResourcePool.prototype.loadResourcesFromFolder = function(folder,recursive) {
	Polycode.ResourcePool_loadResourcesFromFolder(this.__ptr, folder, recursive)
}

ResourcePool.prototype.loadResourcesFromMaterialFile = function(path) {
	Polycode.ResourcePool_loadResourcesFromMaterialFile(this.__ptr, path)
}

ResourcePool.prototype.loadResource = function(path) {
	var retVal = new Resource("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_loadResource(this.__ptr, path)
	return retVal
}

ResourcePool.prototype.loadResourceWithName = function(path,name) {
	var retVal = new Resource("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_loadResourceWithName(this.__ptr, path, name)
	return retVal
}

ResourcePool.prototype.getResource = function(resourceType,resourceName) {
	var retVal = new Resource("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_getResource(this.__ptr, resourceType, resourceName)
	return retVal
}

ResourcePool.prototype.getFont = function(name) {
	var retVal = new Font("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_getFont(this.__ptr, name)
	return retVal
}

ResourcePool.prototype.getMaterial = function(name) {
	var retVal = new Material("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_getMaterial(this.__ptr, name)
	return retVal
}

ResourcePool.prototype.getShader = function(name) {
	var retVal = new Shader("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_getShader(this.__ptr, name)
	return retVal
}

ResourcePool.prototype.loadTexture = function(name) {
	var retVal = new Texture("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_loadTexture(this.__ptr, name)
	return retVal
}

ResourcePool.prototype.loadMesh = function(name) {
	var retVal = new Mesh("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_loadMesh(this.__ptr, name)
	return retVal
}

ResourcePool.prototype.getName = function() {
	return Polycode.ResourcePool_getName(this.__ptr)
}

ResourcePool.prototype.setName = function(name) {
	Polycode.ResourcePool_setName(this.__ptr, name)
}

ResourcePool.prototype.loadFont = function(name,path) {
	var retVal = new Font("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_loadFont(this.__ptr, name, path)
	return retVal
}

ResourcePool.prototype.getResourceByPath = function(resourcePath) {
	var retVal = new Resource("__skip_ptr__")
	retVal.__ptr = Polycode.ResourcePool_getResourceByPath(this.__ptr, resourcePath)
	return retVal
}

ResourcePool.prototype.Update = function(elapsed) {
	Polycode.ResourcePool_Update(this.__ptr, elapsed)
}

ResourcePool.prototype.getResources = function(resourceType) {
	Polycode.ResourcePool_getResources(this.__ptr, resourceType)
}

ResourcePool.prototype.checkForChangedFiles = function() {
	Polycode.ResourcePool_checkForChangedFiles(this.__ptr)
}
